### DISCUSSION TITLE
Password reset flow

### DISCUSSION THREAD

**Discussion by vlaadoo** [supabase/supabase#30402] [Questions] [answered]
How do I track that a user hasn't changed their password after sending a password reset link?

When I reset the password via `supabase.auth.resetPasswordForEmail(email);`, the user has a `recovery_token`

![image](https://github.com/user-attachments/assets/d2c94433-b820-4d34-a657-65091d825fdb)

But as soon as the user clicks on the link from the email, this token disappears.

![image](https://github.com/user-attachments/assets/5acc6e56-c238-4219-a7bc-4ee77073a3c9)

How can I track that the user has not changed his password and redirect him to the password reset page if he tries to access other pages?

Or is it better to somehow handle the password change in such a way that the user is not logged into the account?

My password recovery email template:

```
<h2>Reset Password</h2>

<p>Follow this link to reset the password for your user:</p>
<p>
  <a href="{{ .SiteURL }}/auth/confirm?token_hash={{ .TokenHash}}&type=recovery&next=/account/password-reset">
    Reset Password
  </a>
</p>
```

There was an idea to process this through a password-recovery event, but for some reason, when I click on the link in email, I get a "SIGNED_IN" session ([docs](https://supabase.com/docs/reference/javascript/auth-onauthstatechange))

**Comment by bytaesu** [comment]
I handle this flow in my Next.js 14 (or 15) environment as follows:

### 1. ForgotPassword Page (Not-authenticated)
I call `supabase.auth.resetPasswordForEmail` method with a server action.

### 2. Verification Page (Not-authenticated)
Once the email template is sent, the OTP (Token) is verified on a dedicated verification page using the `supabase.auth.verifyOtp` method within a server action to check its validity.

### 3. ResetPassword Page (Authenticated)
- If the OTP verification is successful, the password change page calls `supabase.auth.updateUser` through a server action to update the password.
- If the user presses the ‘Back’ button to explicitly cancel the flow, they are logged out and redirected to the login page. This is because the updateUser method can only be called when the user is logged in. Logging the user out ensures that if they choose to cancel the password change, they can retry the authentication flow from the beginning.

---

Looking at the API endpoint path, I guess you are using Next.js, so I will attach the repository I wrote in the hope that it will be helpful. I have read a lot of official documentation, including Supabase and Next.js, for clean authentication, so feel free to reach out if you have any further questions or need assistance.

[GitHub Repository: Next.js Supabase Auth Kit](https://github.com/bytaesu/nextjs-supabase-auth-kit)

**Reply by vlaadoo** [reply, marked as answer]
That sounds like what I was looking for, thank you
But your way doesn't quite provide what I need either. If I force type “/dashboard” in the address bar, the app still skips to the main dashboard. 
For now, I've come up with the idea of adding an additional attribute to `raw_user_meta_data` that will severely restrict the user without changing the password

**Reply by bytaesu** [reply]
I get what you're thinking. From what I understand, you seem to be concerned about handling cases where the user navigates away by manipulating the URL directly on the password reset page.

I plan to keep my current approach because I intend to use the ResetPassword page not only for users before login but also for logged-in users who wish to change their passwords. 
Additionally, since users who access the ResetPassword page are already properly authenticated, I don’t think it’s necessary to enforce restrictions on users navigating away by modifying the URL.

I think the method you mentioned will work. Using useEffect to force a logout when the page unmounts, thereby prompting the user to log in again with the new password, could be a simple solution.

I think it's a difference in how you handle the flow, and I hope you find the best way.
